File: bike_route_planner.go
// bike_route_planner.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"crypto/md5"
	"encoding/hex"
)

type Point struct {
	Lat  float64 `json:"lat"`
	Lng  float64 `json:"lng"`
	Name string  `json:"name"`
}

type Route struct {
	Start          Point    `json:"start"`
	End            Point    `json:"end"`
	Waypoints      []Point  `json:"waypoints"`
	TotalDistanceKm float64 `json:"total_distance_km"`
	ElevationGainM int     `json:"elevation_gain_m"`
	TimeHours      float64 `json:"time_hours"`
	Profile        string  `json:"profile"`
	Points         []Point `json:"-"`
}

type Planner struct {
	start     Point
	end       Point
	waypoints []Point
	profile   string
	color     bool
	route     *Route
}

func NewPlanner(start, end string, waypoints []string, profile string, color bool) *Planner {
	return &Planner{
		start:     parseLocation(start),
		end:       parseLocation(end),
		waypoints: parseWaypoints(waypoints),
		profile:   profile,
		color:     color,
	}
}

func parseLocation(loc string) Point {
	if strings.Contains(loc, ",") {
		parts := strings.Split(loc, ",")
		if lat, err1 := strconv.ParseFloat(strings.TrimSpace(parts[0]), 64); err1 == nil {
			if lng, err2 := strconv.ParseFloat(strings.TrimSpace(parts[1]), 64); err2 == nil {
				return Point{Lat: lat, Lng: lng, Name: loc}
			}
		}
	}
	return Point{Name: loc, Lat: 0, Lng: 0}
}

func parseWaypoints(ws []string) []Point {
	var pts []Point
	for _, w := range ws {
		pts = append(pts, parseLocation(w))
	}
	return pts
}

func geocode(name string) Point {
	hash := md5.Sum([]byte(name))
	h := int(hash[0])<<24 | int(hash[1])<<16 | int(hash[2])<<8 | int(hash[3])
	h = h % 360
	return Point{
		Lat:  50 + float64(h%100)/100,
		Lng:  10 + float64(h%200)/100,
		Name: name,
	}
}

func haversine(lat1, lon1, lat2, lon2 float64) float64 {
	const R = 6371.0
	phi1 := lat1 * math.Pi / 180
	phi2 := lat2 * math.Pi / 180
	dphi := (lat2 - lat1) * math.Pi / 180
	dlambda := (lon2 - lon1) * math.Pi / 180
	a := math.Sin(dphi/2)*math.Sin(dphi/2) + math.Cos(phi1)*math.Cos(phi2)*math.Sin(dlambda/2)*math.Sin(dlambda/2)
	return R * 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
}

func getSpeed(profile string) float64 {
	switch profile {
	case "road": return 25
	case "mtb": return 15
	default: return 20
	}
}

func (p *Planner) CalculateRoute() {
	if p.start.Lat == 0 { p.start = geocode(p.start.Name) }
	if p.end.Lat == 0 { p.end = geocode(p.end.Name) }
	for i := range p.waypoints {
		if p.waypoints[i].Lat == 0 {
			p.waypoints[i] = geocode(p.waypoints[i].Name)
		}
	}
	points := []Point{p.start}
	points = append(points, p.waypoints...)
	points = append(points, p.end)

	var totalDist float64
	for i := 0; i < len(points)-1; i++ {
		totalDist += haversine(points[i].Lat, points[i].Lng, points[i+1].Lat, points[i+1].Lng)
	}
	speed := getSpeed(p.profile)
	timeHours := totalDist / speed
	elevGain := int(totalDist * 0.05 * 1000)

	p.route = &Route{
		Start:          p.start,
		End:            p.end,
		Waypoints:      p.waypoints,
		TotalDistanceKm: totalDist,
		ElevationGainM: elevGain,
		TimeHours:      timeHours,
		Profile:        p.profile,
		Points:         points,
	}
}

func (p *Planner) PrintRoute() {
	if p.route == nil { p.CalculateRoute() }
	r := p.route
	if p.color {
		fmt.Printf("\033[36m🚴 Маршрут от %s до %s\033[0m\n", p.start.Name, p.end.Name)
		if len(p.waypoints) > 0 {
			names := make([]string, len(p.waypoints))
			for i, w := range p.waypoints { names[i] = w.Name }
			fmt.Printf("\033[35m   через %s\033[0m\n", strings.Join(names, ", "))
		}
		fmt.Printf("\033[32mРасстояние: %.1f км\033[0m\n", r.TotalDistanceKm)
		fmt.Printf("\033[33mНабор высоты: %d м\033[0m\n", r.ElevationGainM)
		fmt.Printf("\033[34mВремя в пути: ~%.1f ч\033[0m\n", r.TimeHours)
	} else {
		fmt.Printf("Маршрут от %s до %s\n", p.start.Name, p.end.Name)
		if len(p.waypoints) > 0 {
			names := make([]string, len(p.waypoints))
			for i, w := range p.waypoints { names[i] = w.Name }
			fmt.Printf("   через %s\n", strings.Join(names, ", "))
		}
		fmt.Printf("Расстояние: %.1f км\n", r.TotalDistanceKm)
		fmt.Printf("Набор высоты: %d м\n", r.ElevationGainM)
		fmt.Printf("Время в пути: ~%.1f ч\n", r.TimeHours)
	}
}

func (p *Planner) ExportGPX(filename string) {
	if p.route == nil { p.CalculateRoute() }
	gpx := `<?xml version="1.0" encoding="UTF-8"?>
<gpx version="1.1" creator="BikeRoutePlanner">
  <trk>
    <trkseg>
`
	for _, pt := range p.route.Points {
		gpx += fmt.Sprintf(`      <trkpt lat="%.6f" lon="%.6f">
        <ele>%d</ele>
      </trkpt>
`, pt.Lat, pt.Lng, int(50+pt.Lat*10))
	}
	gpx += `    </trkseg>
  </trk>
</gpx>`
	os.WriteFile(filename, []byte(gpx), 0644)
	fmt.Printf("GPX экспортирован в %s\n", filename)
}

func (p *Planner) SaveJSON(filename string) {
	if p.route == nil { p.CalculateRoute() }
	data, _ := json.MarshalIndent(p.route, "", "  ")
	os.WriteFile(filename, data, 0644)
	fmt.Printf("Маршрут сохранён в %s\n", filename)
}

func main() {
	var (
		start     string
		end       string
		waypoints string
		profile   string
		gpx       string
		jsonOut   string
		color     bool
	)
	flag.StringVar(&start, "start", "", "Начальная точка")
	flag.StringVar(&end, "end", "", "Конечная точка")
	flag.StringVar(&waypoints, "waypoints", "", "Промежуточные точки через запятую")
	flag.StringVar(&profile, "profile", "touring", "road, mtb, touring")
	flag.StringVar(&gpx, "export-gpx", "", "Экспорт в GPX")
	flag.StringVar(&jsonOut, "save-json", "", "Сохранить в JSON")
	flag.BoolVar(&color, "color", false, "Принудительный цветной вывод")
	flag.Parse()

	if start == "" || end == "" {
		fmt.Fprintln(os.Stderr, "Error: --start and --end are required")
		os.Exit(1)
	}
	var ws []string
	if waypoints != "" {
		ws = strings.Split(waypoints, ",")
	}
	planner := NewPlanner(start, end, ws, profile, color || isTerminal())
	planner.CalculateRoute()
	planner.PrintRoute()
	if gpx != "" { planner.ExportGPX(gpx) }
	if jsonOut != "" { planner.SaveJSON(jsonOut) }
}

func isTerminal() bool {
	stat, _ := os.Stdout.Stat()
	return (stat.Mode() & os.ModeCharDevice) != 0
}

File: bike_route_planner.rs
// bike_route_planner.rs
use clap::{App, Arg};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::f64::consts::PI;
use colored::*;

#[derive(Debug, Clone, Serialize, Deserialize)]
struct Point {
    lat: f64,
    lng: f64,
    name: String,
}

#[derive(Debug, Serialize)]
struct Route {
    start: Point,
    end: Point,
    waypoints: Vec<Point>,
    total_distance_km: f64,
    elevation_gain_m: i32,
    time_hours: f64,
    profile: String,
    points: Vec<Point>,
}

struct Planner {
    start: Point,
    end: Point,
    waypoints: Vec<Point>,
    profile: String,
    color: bool,
    route: Option<Route>,
}

impl Planner {
    fn new(start: &str, end: &str, waypoints: Vec<&str>, profile: &str, color: bool) -> Self {
        Planner {
            start: Self::parse_location(start),
            end: Self::parse_location(end),
            waypoints: waypoints.into_iter().map(Self::parse_location).collect(),
            profile: profile.to_string(),
            color,
            route: None,
        }
    }

    fn parse_location(loc: &str) -> Point {
        if let Some((lat_str, lng_str)) = loc.split_once(',') {
            if let (Ok(lat), Ok(lng)) = (lat_str.trim().parse(), lng_str.trim().parse()) {
                return Point { lat, lng, name: loc.to_string() };
            }
        }
        Point { lat: 0.0, lng: 0.0, name: loc.to_string() }
    }

    fn geocode(name: &str) -> Point {
        use std::collections::hash_map::DefaultHasher;
        use std::hash::{Hash, Hasher};
        let mut hasher = DefaultHasher::new();
        name.hash(&mut hasher);
        let h = hasher.finish() % 360;
        Point {
            lat: 50.0 + (h % 100) as f64 / 100.0,
            lng: 10.0 + (h % 200) as f64 / 100.0,
            name: name.to_string(),
        }
    }

    fn haversine(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
        const R: f64 = 6371.0;
        let phi1 = lat1 * PI / 180.0;
        let phi2 = lat2 * PI / 180.0;
        let dphi = (lat2 - lat1) * PI / 180.0;
        let dlambda = (lon2 - lon1) * PI / 180.0;
        let a = (dphi/2.0).sin().powi(2) + phi1.cos() * phi2.cos() * (dlambda/2.0).sin().powi(2);
        R * 2.0 * a.sqrt().atan2((1.0 - a).sqrt())
    }

    fn get_speed(profile: &str) -> f64 {
        match profile {
            "road" => 25.0,
            "mtb" => 15.0,
            _ => 20.0,
        }
    }

    fn calculate_route(&mut self) {
        if self.start.lat == 0.0 { self.start = Self::geocode(&self.start.name); }
        if self.end.lat == 0.0 { self.end = Self::geocode(&self.end.name); }
        for w in &mut self.waypoints {
            if w.lat == 0.0 { *w = Self::geocode(&w.name); }
        }
        let mut points = vec![self.start.clone()];
        points.extend(self.waypoints.clone());
        points.push(self.end.clone());

        let mut total_dist = 0.0;
        for i in 0..points.len()-1 {
            total_dist += Self::haversine(points[i].lat, points[i].lng, points[i+1].lat, points[i+1].lng);
        }
        let speed = Self::get_speed(&self.profile);
        let time_hours = total_dist / speed;
        let elev_gain = (total_dist * 0.05 * 1000.0) as i32;

        self.route = Some(Route {
            start: self.start.clone(),
            end: self.end.clone(),
            waypoints: self.waypoints.clone(),
            total_distance_km: total_dist,
            elevation_gain_m: elev_gain,
            time_hours,
            profile: self.profile.clone(),
            points,
        });
    }

    fn print_route(&self) {
        let r = self.route.as_ref().unwrap();
        if self.color {
            println!("{}", format!("🚴 Маршрут от {} до {}", self.start.name, self.end.name).cyan());
            if !self.waypoints.is_empty() {
                let names: Vec<_> = self.waypoints.iter().map(|w| w.name.as_str()).collect();
                println!("{}", format!("   через {}", names.join(", ")).magenta());
            }
            println!("{}", format!("Расстояние: {:.1} км", r.total_distance_km).green());
            println!("{}", format!("Набор высоты: {} м", r.elevation_gain_m).yellow());
            println!("{}", format!("Время в пути: ~{:.1} ч", r.time_hours).blue());
        } else {
            println!("Маршрут от {} до {}", self.start.name, self.end.name);
            if !self.waypoints.is_empty() {
                let names: Vec<_> = self.waypoints.iter().map(|w| w.name.as_str()).collect();
                println!("   через {}", names.join(", "));
            }
            println!("Расстояние: {:.1} км", r.total_distance_km);
            println!("Набор высоты: {} м", r.elevation_gain_m);
            println!("Время в пути: ~{:.1} ч", r.time_hours);
        }
    }

    fn export_gpx(&self, filename: &str) {
        let r = self.route.as_ref().unwrap();
        let mut gpx = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n".to_string();
        gpx.push_str("<gpx version=\"1.1\" creator=\"BikeRoutePlanner\">\n");
        gpx.push_str("  <trk>\n    <trkseg>\n");
        for pt in &r.points {
            gpx.push_str(&format!("      <trkpt lat=\"{:.6}\" lon=\"{:.6}\">\n", pt.lat, pt.lng));
            gpx.push_str(&format!("        <ele>{}</ele>\n", (50.0 + pt.lat * 10.0) as i32));
            gpx.push_str("      </trkpt>\n");
        }
        gpx.push_str("    </trkseg>\n  </trk>\n</gpx>");
        fs::write(filename, gpx).unwrap();
        println!("GPX экспортирован в {}", filename);
    }

    fn save_json(&self, filename: &str) {
        let json = serde_json::to_string_pretty(&self.route.as_ref().unwrap()).unwrap();
        fs::write(filename, json).unwrap();
        println!("Маршрут сохранён в {}", filename);
    }
}

fn main() {
    let matches = App::new("Bike Route Planner")
        .arg(Arg::with_name("start").long("start").takes_value(true).required(true))
        .arg(Arg::with_name("end").long("end").takes_value(true).required(true))
        .arg(Arg::with_name("waypoints").long("waypoints").takes_value(true))
        .arg(Arg::with_name("profile").long("profile").takes_value(true).default_value("touring"))
        .arg(Arg::with_name("export-gpx").long("export-gpx").takes_value(true))
        .arg(Arg::with_name("save-json").long("save-json").takes_value(true))
        .arg(Arg::with_name("color").long("color"))
        .get_matches();

    let start = matches.value_of("start").unwrap();
    let end = matches.value_of("end").unwrap();
    let waypoints = matches.value_of("waypoints").map(|s| s.split(',').collect()).unwrap_or(vec![]);
    let profile = matches.value_of("profile").unwrap();
    let color = matches.is_present("color") || atty::is(atty::Stream::Stdout);

    let mut planner = Planner::new(start, end, waypoints, profile, color);
    planner.calculate_route();
    planner.print_route();
    if let Some(gpx) = matches.value_of("export-gpx") { planner.export_gpx(gpx); }
    if let Some(json) = matches.value_of("save-json") { planner.save_json(json); }
}
